go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+	SetBitsBuffer(*memory.Buffer)
+}
+
+func computeLevelInfo(descr *schema.Column) (info LevelInfo) {
+	info.DefLevel = descr.MaxDefinitionLevel()
+	info.RepLevel = descr.MaxRepetitionLevel()
+
+	minSpacedDefLevel := descr.MaxDefinitionLevel()
+	n := descr.SchemaNode()
+	for n != nil && n.RepetitionType() != parquet.Repetitions.Repeated {
+		if n.RepetitionType() == parquet.Repetitions.Optional {
+			minSpacedDefLevel--
+		}
+		n = n.Parent()
+	}
+	info.RepeatedAncestorDefLevel = minSpacedDefLevel
+	return
+}
+
+type columnWriter struct {
+	metaData *metadata.ColumnChunkMetaDataBuilder
+	descr    *schema.Column
+
+	// scratch buffer if validity bits need to be recalculated
+	bitsBuffer *memory.Buffer
+	levelInfo  LevelInfo
+	pager      PageWriter
+	hasDict    bool
+	encoding   parquet.Encoding
+	props      *parquet.WriterProperties
+	defEncoder encoding.LevelEncoder
+	repEncoder encoding.LevelEncoder
+	mem        memory.Allocator
+
+	pageStatistics  metadata.TypedStatistics
+	chunkStatistics metadata.TypedStatistics
+
+	// total number of values stored in the data page. this is the maximum
+	// of the number of encoded def levels or encoded values. for
+	// non-repeated, required columns, this is equal to the number of encoded
+	// values. For repeated or optional values, there may be fewer data values
+	// than levels, and this tells you how many encoded levels there are in that case
+	numBuffered int64
+
+	// the total number of stored values. for repeated or optional values. this
+	// number may be lower than numBuffered
+	numBufferedEncoded int64
+
+	rowsWritten       int
+	totalBytesWritten int64
+	// records the current number of compressed bytes in a column
+	totalCompressedBytes int64
+	closed               bool
+	fallback             bool
+
+	pages []DataPage
+
+	defLevelSink *encoding.PooledBufferWriter
+	repLevelSink *encoding.PooledBufferWriter
+
+	uncompressedData bytes.Buffer
+	compressedTemp   *bytes.Buffer
+
+	currentEncoder encoding.TypedEncoder
+}
+
+func newColumnWriterBase(metaData *metadata.ColumnChunkMetaDataBuilder, pager PageWriter, useDict bool, enc parquet.Encoding, props *parquet.WriterProperties) columnWriter {
+	ret := columnWriter{
+		metaData:     metaData,
+		descr:        metaData.Descr(),
+		levelInfo:    computeLevelInfo(metaData.Descr()),
+		pager:        pager,
+		hasDict:      useDict,
+		encoding:     enc,
+		props:        props,
+		mem:          props.Allocator(),
+		defLevelSink: encoding.NewPooledBufferWriter(0),
+		repLevelSink: encoding.NewPooledBufferWriter(0),
+	}
+	if pager.HasCompressor() {
+		ret.compressedTemp = new(bytes.Buffer)
+	}
+	if props.StatisticsEnabledFor(ret.descr.Path()) && ret.descr.SortOrder() != schema.SortUNKNOWN {
+		ret.pageStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+		ret.chunkStatistics = metadata.NewStatistics(ret.descr, props.Allocator())
+	}
+
+	if ret.props.DataPageVersion() == parquet.DataPageV1 {
+		if ret.descr.MaxDefinitionLevel() > 0 {
```

Comment:
maybe add a comment on why the offset is being added (this is so the number of levels can be recorded at the beginning of the buffer?)

Reply:
comment added